nio/block/mixins/retry/retry.py

```diff
@@ -191,3 +191,14 @@ def before_retry(self, *args, **kwargs):
         any other behavior that may make the next retry attempt succeed.
         """
         pass  # pragma: no cover
+
+    def max_retry_exhausted(self, exc):
```

Comment:
Interesting idea to allow different behavior when we're done retrying, I like it!